File: test/tp_drafter_catalog_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:psyclinicai/services/assessments/outcome_measure_catalog.dart';
import 'package:psyclinicai/services/treatment_plan_drafter/tp_drafter_catalog.dart';

void main() {
  group('TpDrafterCatalog invariants', () {
    test('non-empty + unique (disorder, modality) keys', () {
      expect(TpDrafterCatalog.protocols, isNotEmpty);
      final seen = <String>{};
      for (final p in TpDrafterCatalog.protocols) {
        final key = '${p.disorder.name}_${p.modality.name}';
        expect(seen.contains(key), isFalse,
            reason: '$key: duplicate');
        seen.add(key);
      }
    });

    test('every protocol pins at least one guideline anchor', () {
      for (final p in TpDrafterCatalog.protocols) {
        expect(p.guidelineAnchors, isNotEmpty,
            reason: '${p.label}: missing guideline anchor');
      }
    });

    test('outcomeInstrument refers to a known outcome measure', () {
      for (final p in TpDrafterCatalog.protocols) {
        final m = OutcomeMeasureCatalog.byScaleId(p.outcomeInstrument);
        expect(m, isNotNull,
            reason:
                '${p.label}: outcomeInstrument ${p.outcomeInstrument} unknown');
      }
    });

    test('PTSD + BPD + AUD protocols require supervisor co-sign', () {
      for (final p in TpDrafterCatalog.protocols) {
        final highRisk = p.disorder == TpDisorderId.ptsd ||
            p.disorder == TpDisorderId.borderlinePersonalityDisorder ||
            p.disorder == TpDisorderId.alcoholUseDisorder;
        if (highRisk) {
          expect(p.requiresSupervisorCoSign, isTrue,
              reason: '${p.label}: high-risk modality must require co-sign');
        }
      }
    });

    test('recommendedSessions positive', () {
      for (final p in TpDrafterCatalog.protocols) {
        expect(p.recommendedSessions > 0, isTrue);
      }
    });

    test('byKey returns matching protocol', () {
      final p = TpDrafterCatalog.byKey(
        disorder: TpDisorderId.majorDepressiveDisorder,
        modality: TpModality.cbt,
      );
      expect(p.label, contains('Major Depressive Disorder'));
    });

    test('byKey throws on unsupported tuple', () {
      expect(
        () => TpDrafterCatalog.byKey(
          disorder: TpDisorderId.insomniaDisorder,
          modality: TpModality.emdr,
        ),
        throwsStateError,
      );
    });

    test('modalitiesFor returns non-empty for every disorder we ship', () {
      for (final p in TpDrafterCatalog.protocols) {
        final mods = TpDrafterCatalog.modalitiesFor(p.disorder);
        expect(mods, isNotEmpty);
        expect(mods.contains(p.modality), isTrue);
      }
    });

    test('smartGoalFields + outputSections non-empty + unique', () {
      expect(TpDrafterCatalog.smartGoalFields, isNotEmpty);
      expect(
        TpDrafterCatalog.smartGoalFields.toSet().length,
        TpDrafterCatalog.smartGoalFields.length,
      );
      expect(TpDrafterCatalog.outputSections, isNotEmpty);
      expect(
        TpDrafterCatalog.outputSections.toSet().length,
        TpDrafterCatalog.outputSections.length,
      );
    });

    test('schemaVersion positive + lastReviewed YYYY-MM', () {
      expect(TpDrafterCatalog.schemaVersion > 0, isTrue);
      expect(TpDrafterCatalog.lastReviewed,
          matches(RegExp(r'^\d{4}-\d{2}$')));
    });
  });

  group('requiresCoSign helper', () {
    test('returns true for high-risk PTSD-EMDR', () {
      final spec = TpDrafterCatalog.byKey(
        disorder: TpDisorderId.ptsd,
        modality: TpModality.emdr,
      );
      expect(requiresCoSign(spec), isTrue);
    });

    test('returns false for low-risk MDD-CBT', () {
      final spec = TpDrafterCatalog.byKey(
        disorder: TpDisorderId.majorDepressiveDisorder,
        modality: TpModality.cbt,
      );
      expect(requiresCoSign(spec), isFalse);
    });
  });
}

File: functions/src/__tests__/tp_drafter_catalog.test.ts
import {
  TPD_LAST_REVIEWED,
  TPD_OUTPUT_SECTIONS,
  TPD_PROTOCOLS,
  TPD_SCHEMA_VERSION,
  TPD_SMART_GOAL_FIELDS,
  jsonSchemaForPlan,
  tpProtocolByKey,
} from "../lib/tp_drafter_catalog";

describe("TPD catalog invariants", () => {
  test("non-empty + unique (disorder, modality) tuples", () => {
    expect(TPD_PROTOCOLS.length).toBeGreaterThan(0);
    const seen = new Set<string>();
    for (const p of TPD_PROTOCOLS) {
      const k = `${p.disorder}_${p.modality}`;
      expect(seen.has(k)).toBe(false);
      seen.add(k);
    }
  });

  test("every protocol has guideline anchors", () => {
    for (const p of TPD_PROTOCOLS) {
      expect(p.guidelineAnchors.length).toBeGreaterThan(0);
    }
  });

  test("high-risk protocols require supervisor co-sign", () => {
    for (const p of TPD_PROTOCOLS) {
      if (
        p.disorder === "ptsd" ||
        p.disorder === "borderlinePersonalityDisorder" ||
        p.disorder === "alcoholUseDisorder"
      ) {
        expect(p.requiresSupervisorCoSign).toBe(true);
      }
    }
  });

  test("schemaVersion + lastReviewed shape", () => {
    expect(TPD_SCHEMA_VERSION).toBeGreaterThan(0);
    expect(TPD_LAST_REVIEWED).toMatch(/^\d{4}-\d{2}$/);
  });

  test("smart goal fields + sections non-empty + unique", () => {
    expect(TPD_SMART_GOAL_FIELDS.length).toBeGreaterThan(0);
    expect(new Set(TPD_SMART_GOAL_FIELDS).size).toBe(
      TPD_SMART_GOAL_FIELDS.length,
    );
    expect(TPD_OUTPUT_SECTIONS.length).toBeGreaterThan(0);
    expect(new Set(TPD_OUTPUT_SECTIONS).size).toBe(
      TPD_OUTPUT_SECTIONS.length,
    );
  });
});

describe("tpProtocolByKey", () => {
  test("returns matching", () => {
    const p = tpProtocolByKey({
      disorder: "majorDepressiveDisorder",
      modality: "cbt",
    });
    expect(p.recommendedSessions).toBe(16);
  });

  test("throws on unsupported tuple", () => {
    expect(() =>
      tpProtocolByKey({disorder: "insomniaDisorder", modality: "emdr"}),
    ).toThrow();
  });
});

describe("jsonSchemaForPlan", () => {
  test("required list matches output sections", () => {
    const schema = jsonSchemaForPlan() as {
      required: string[];
      properties: Record<string, unknown>;
      additionalProperties: boolean;
    };
    expect(schema.required.sort()).toEqual(
      [...TPD_OUTPUT_SECTIONS].sort(),
    );
    expect(schema.additionalProperties).toBe(false);
  });

  test("smart_goals items require every SMART field", () => {
    const schema = jsonSchemaForPlan() as {
      properties: {
        smart_goals: {items: {required: string[]}};
      };
    };
    expect(
      schema.properties.smart_goals.items.required.sort(),
    ).toEqual([...TPD_SMART_GOAL_FIELDS].sort());
  });
});
